File: psra/service/src/main/java/by/arhor/psra/service/impl/UserServiceImpl.java
package by.arhor.psra.service.impl;

import by.arhor.psra.dto.UserDto;
import by.arhor.psra.exception.EntityDuplicateException;
import by.arhor.psra.exception.EntityNotFoundException;
import by.arhor.psra.localization.ErrorLabel;
import by.arhor.psra.model.User;
import by.arhor.psra.repository.UserRepository;
import by.arhor.psra.service.UserService;
import by.arhor.psra.util.MappingDelegate;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Primary;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.function.Predicate;

import static by.arhor.psra.localization.ErrorLabel.USER_DUPLICATE;
import static by.arhor.psra.localization.ErrorLabel.USER_NOT_FOUND;
import static java.util.Collections.singletonList;

@Primary
@Service
@Transactional
public class UserServiceImpl
    extends AbstractService<User, UserDto, String>
    implements UserService
             , UserDetailsService {

  private final PasswordEncoder encoder;

  @Autowired
  public UserServiceImpl(UserRepository repository,
                         PasswordEncoder encoder,
                         MappingDelegate mapper) {
    super(repository, mapper, UserDto.class);
    this.encoder = encoder;
  }

  @Override
  protected ErrorLabel notFoundLabel() {
    return USER_NOT_FOUND;
  }

  @Override
	@Transactional(readOnly = true)
	public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
    return ((UserRepository) repository)
        .findByUsername(username)
        .map(user ->
            new org.springframework.security.core.userdetails.User(
                user.getUsername(),
                user.getPassword(),
                singletonList(new SimpleGrantedAuthority(user.getRole().toString()))))
			  .orElseThrow(() -> new UsernameNotFoundException(username));
  }

  @Override
	public UserDto create(UserDto dto) {
    checkUsername(dto.getUsername());
    checkEmail(dto.getEmail());

		final var user = mapper.map(dto, User.class);

		user.setPassword(encoder.encode(user.getPassword()));

		final var created = repository.insert(user);
		return mapper.map(created, UserDto.class);
	}

	@Override
	public UserDto update(UserDto dto) {
    final var user = repository
        .findById(dto.getId())
        .orElseThrow(() -> new EntityNotFoundException(USER_NOT_FOUND, "ID", dto.getId()));

    user.setUsername(dto.getUsername());
    user.setPassword(encoder.encode(dto.getPassword()));

    final var updated = repository.save(user);

    return mapper.map(updated, UserDto.class);
  }

  private void checkUsername(String username) {
    checkForDuplicates(((UserRepository) repository)::existsByUsername, username, "Username");
  }

  private void checkEmail(String email) {
    checkForDuplicates(((UserRepository) repository)::existsByEmail, email, "Email");
  }

  private <T> void checkForDuplicates(Predicate<T> checker, T value, String fieldName) {
    final var isTaken = checker.test(value);
    if (isTaken) {
      throw new EntityDuplicateException(USER_DUPLICATE, fieldName, value);
    }
  }
}

File: psra/service/src/main/scala/by/arhor/psra/service/impl/UserServiceImpl.scala
package by.arhor.psra.service.impl

import java.util
import java.util.Collections
import java.util.stream.Collectors.toList

import by.arhor.psra.dto.UserDto
import by.arhor.psra.exception.EntityNotFoundException
import by.arhor.psra.localization.ErrorLabel
import by.arhor.psra.model.User
import by.arhor.psra.repository.{CourseRepository, UserRepository}
import by.arhor.psra.service.UserService
import org.modelmapper.ModelMapper
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.annotation.Primary
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.{UserDetails, UserDetailsService, UsernameNotFoundException}
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Primary
@Service
@Transactional
class UserServiceImpl @Autowired() (
  private val repository: UserRepository,
	private val courseRepository: CourseRepository,
	private val encoder: PasswordEncoder,
	override protected val modelMapper: ModelMapper
) extends UserService
     with UserDetailsService {

	type SpringUser = org.springframework.security.core.userdetails.User

	@throws(classOf[UsernameNotFoundException])
	@Transactional(readOnly = true)
	override def loadUserByUsername(username: String): UserDetails =
		repository
			.findByUsername(username)
  		.map[UserDetails] { user =>
				new SpringUser(
					user.username,
					user.password,
					Collections.singletonList(new SimpleGrantedAuthority(user.role.toString))
				)
			}
			.orElseThrow { () => new UsernameNotFoundException(username) }

	@Transactional(readOnly = true)
	override def findOne(id: String): UserDto =
		repository
			.findById(id)
			.map[UserDto] { mapToDto }
  		.orElseThrow {
				() => new EntityNotFoundException(ErrorLabel.USER_NOT_FOUND, "ID", id)
			}
	
	@Transactional(readOnly = true)
	override def findAll(): util.List[UserDto] =
		repository
			.findAll
			.stream
			.map[UserDto] { mapToDto }
			.collect(toList())

	override def findLearnersByCourseId(cid: String): util.List[UserDto] =
		courseRepository
			.findById(cid)
			.map[util.List[UserDto]] { course =>
				course
					.learners
					.stream
					.map[UserDto] { mapToDto }
					.collect(toList())
			}
			.orElseThrow {
				() => new EntityNotFoundException(ErrorLabel.COURSE_NOT_FOUND, "ID", cid)
			}

	override def create(dto: UserDto): UserDto = {
		// FIXME Check for name duplicates

		lazy val user: User = mapToEntity(dto) // lazy?
		user.password = encoder.encode(user.password)
		lazy val created: User = repository.insert(user) // lazy?
		mapToDto(created)
	}

	override def update(dto: UserDto): UserDto = ??? // TODO Auto-generated method stub

	override def delete(dto: UserDto): Unit =
		repository
			.findById(dto.getId)
			.map[Unit] { repository delete _ } // TODO: does it work?
			.orElseThrow {
				() => new EntityNotFoundException(ErrorLabel.USER_NOT_FOUND, "ID", dto.getId)
			}
}
